### PR TITLE
[ProductCatalogService] Add product catalog service problem patterns

### PR DESCRIPTION
**Pull request by klucsik**
# Changes

Refactored the featre flag querying in ProductCatalogService to be reusable
Added Problems to ListProducts and GetProduct endpoints in the product catalog service:

- ItemListInternal - will fail ListProducts with Internal error code
- ItemNotFound - will fail GetProduct with NotFound error code

## Merge Requirements

For new features contributions please make sure you have completed the following
essential items:

* [x] `CHANGELOG.md` updated to document new feature additions
* [x] Appropriate documentation updates in the [docs](https://github.com/open-telemetry/opentelemetry.io/pull/4004)
~~* [ ] Appropriate Helm chart updates in the [helm-charts][]~~

### CODE REVIEW

**Comment by julianocosta89** [comment]
Hey @klucsik, nice to see you here!
Before we check your PR, could you sign the EasyCLA?

**Comment by klucsik** [comment]
> Hey @klucsik, nice to see you here! Before we check your PR, could you sign the EasyCLA?

Yes yes, we are on it, just having some learning path with commit signing :D

**Comment by klucsik** [comment]
> Hey @klucsik, nice to see you here! Before we check your PR, could you sign the EasyCLA?

Nice to work with you again :) EasyCLA is green

File: src/ffspostgres/init-scripts/20-ffs_data.sql [hunk 1]
```diff
@@ -7,4 +7,6 @@ VALUES
     ('productCatalogFailure', 'Fail product catalog service on a specific product', 0),
     ('recommendationCache', 'Cache recommendations', 0),
     ('adServiceFailure', 'Fail ad service requests', 0),
-    ('cartServiceFailure', 'Fail cart service requests', 0);
+    ('cartServiceFailure', 'Fail cart service requests', 0)
```
Comment (julianocosta89) [inline src/ffspostgres/init-scripts/20-ffs_data.sql:10]:
```suggestion
    ('cartServiceFailure', 'Fail cart service requests', 0),
```

File: src/productcatalogservice/main.go [hunk 2]
```diff
@@ -237,6 +244,13 @@ func (p *productCatalog) GetProduct(ctx context.Context, req *pb.GetProductReque
 		return nil, status.Errorf(codes.Internal, msg)
 	}
 
+	if p.getFeatureFlag(ctx, "ItemNotFound") {
```
[the remark this answers is not in the document]
Comment (julianocosta89) [reply src/productcatalogservice/main.go:247]: The flag would still need to be in place to control when to send the wrong id, but yeah, agree with you.

[the remark this answers is not in the document]
Comment (julianocosta89) [reply src/productcatalogservice/main.go:247]: Hmmmm, validating the PR I actually understood the scenario.
`ItemNotFound` is being triggered just when navigating to the product page.

I can see that being used in demos, to showcase errors on a page that is part of the checkout flow.


`ItemListInternal` on the other hand, fails on listing the products, so you can't even see the products.

**Comment by puckpuck** [comment]
I agree with @austinlparker's comment about introducing another feature flag for a simple blunt failure scenario. We already have several similar feature flags across different aspects of the demo. What observability demo and/or use cases are we trying to solve with this feature flag?

File: src/productcatalogservice/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (klucsik) [reply src/productcatalogservice/main.go:247]: I could move the ItemNotFound logic to loadgenerator to query some clearly invalid products, the requirement from our end is to demonstrate a notfound error in the sample

**Comment by klucsik** [comment]
@austinlparker @julianocosta89 @puckpuck Thank you for reading through my PR, and the comments.
What do you mean of triggering 100% failure scenarios? As I understand, and experienced, the problem scenarios can be triggered partially if we set the feature flag in the feature flag service like 0.3 (which gets triggered every 30% of the calls).
My goal is basically to introduce 404 and 500 errors in a Go service, so we can demonstrate what can one see with open telemetry data, and with more instrumentation. Initally these meant to be in the SearchProducts, but turned out, nothing calls the SearchProducts endpoint (and the loadgenerator doesnt seem to handle gRPC requests) so I moved them where it made most sense to me.
I could trigger the 404 from the loadgenerator side, but 500 should be in the ProductCatalogService.
I could imagine the ListProduct would be able to error out if it gets a malformed data json, or if we would use some DB to store the products and that would have an outage or similar problem.
I'm very much open for your suggestions, my goal would be to introduce these errortypes in a meaningfull form, triggerable from the featureflag service.

**Comment by puckpuck** [comment]
@klucsik we already have a feature flag in the Product Catalog service that will return a 500 for a specific product ID when the flag is enabled. Would the `productCatalogFailure` feature flag support your need to see a Go service return a 500 under some conditions? https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/productcatalogservice/main.go#L279-L303

I want to leverage feature flags to showcase potential failures, but I would like the failures to be more than just blunt failures on a feature flag being positive. It would help that the failure was due to a specific scenario, such as a product ID or multiple specific products being purchased, etc.

**Comment by julianocosta89** [comment]
@klucsik we have just merged a PR that removed the current FeatureFlagService in favor of OpenFeature and FlagD.
Unfortunately your PR needs to be revisited.
Would you be able to update it?

**Comment by klucsik** [comment]
Hey @julianocosta89. Sorry for going radiosilent about this. We discussed it a lot internally.
I'll definetly rework this, as we need these problem patterns present, but maybe we shouldn't merge it to upstream as you guys pointed out that from pure otel perspective these arent providing much value.

**Comment by julianocosta89** [comment]
@klucsik got it.
Should we close this PR then?

**Pull request closed without merging**